quickfixj-core/src/main/java/quickfix/Session.java

```diff
@@ -2321,71 +2323,85 @@ private void nextLogon(Message logon) throws FieldNotFound, RejectLogon, Incorre
-        try {
-            state.get(beginSeqNo, endSeqNo, messages);
-        } catch (final IOException e) {
-            if (forceResendWhenCorruptedStore) {
-                LOG.error("Cannot read messages from stores, resend HeartBeats", e);
-                for (int i = beginSeqNo; i < endSeqNo; i++) {
-                    final Message heartbeat = messageFactory.create(sessionID.getBeginString(),
-                            MsgType.HEARTBEAT);
-                    initializeHeader(heartbeat.getHeader());
-                    heartbeat.getHeader().setInt(MsgSeqNum.FIELD, i);
-                    messages.add(heartbeat.toString());
-                }
-            } else {
-                throw e;
-            }
-        }
-
         int msgSeqNum = 0;
         int begin = 0;
         int current = beginSeqNo;
         boolean appMessageJustSent = false;
 
-        for (final String message : messages) {
-            appMessageJustSent = false;
-            final Message msg;
+        int curBatchStartSeqNo = beginSeqNo;
+        while (curBatchStartSeqNo <= endSeqNo) {
+            int endCurBatchSeqNo = endSeqNo;
+            if (curBatchStartSeqNo + MAX_RESEND_BATCH_RETRIEVAL_SIZE < endSeqNo) {
+                endCurBatchSeqNo = curBatchStartSeqNo + MAX_RESEND_BATCH_RETRIEVAL_SIZE;
+            }
+            final ArrayList<String> messages = new ArrayList<>();
             try {
-                // QFJ-626
-                msg = parseMessage(message);
-                msgSeqNum = msg.getHeader().getInt(MsgSeqNum.FIELD);
-            } catch (final Exception e) {
-                getLog().onErrorEvent(
-                        "Error handling ResendRequest: failed to parse message (" + e.getMessage()
-                        + "): " + message);
-                // Note: a SequenceReset message will be generated to fill the gap
-                continue;
+                state.get(curBatchStartSeqNo, endCurBatchSeqNo, messages);
+            } catch (final IOException e) {
+                if (forceResendWhenCorruptedStore) {
+                    LOG.error("Cannot read messages from stores, resend HeartBeats", e);
+                    for (int i = beginSeqNo; i < endSeqNo; i++) {
+                        final Message heartbeat = messageFactory.create(sessionID.getBeginString(),
+                            MsgType.HEARTBEAT);
+                        initializeHeader(heartbeat.getHeader());
+                        heartbeat.getHeader().setInt(MsgSeqNum.FIELD, i);
+                        messages.add(heartbeat.toString());
+                    }
+                } else {
+                    throw e;
+                }
             }
+            for (final String message : messages) {
+                appMessageJustSent = false;
+                final Message msg;
+                try {
+                    // QFJ-626
+                    msg = parseMessage(message);
+                    if (msg.getException() != null) {
```

Comment:
@philipwhiuk I am not sure about this. This will not resend messages which were formerly (without this PR) sent. I mean we could discuss if it is sensible but I feel that this should be a separate change. WDYT?